src/hotspot/share/gc/shared/referenceProcessor.cpp

```diff
@@ -633,60 +632,60 @@ void ReferenceProcessor::balance_queues(DiscoveredList ref_lists[])
   size_t avg_refs = total_refs / _num_queues + 1;
   uint to_idx = 0;
   for (uint from_idx = 0; from_idx < _max_num_queues; from_idx++) {
-    bool move_all = false;
+    size_t from_len = ref_lists[from_idx].length();
+
+    size_t remaining_to_move;
     if (from_idx >= _num_queues) {
-      move_all = ref_lists[from_idx].length() > 0;
+      // Move all
+      remaining_to_move = from_len;
+    } else {
+      // Move those above avg_refs
+      remaining_to_move = from_len > avg_refs
+                        ? from_len - avg_refs
+                        : 0;
```

Comment:
Continuing the discussion about this snippet.

I think that use of the ternary operator here looks weird to me, being
combined with the if-statement that way. I might write
```
size_t remaining_to_move = from_len;
if (from_idx < _num_queues) {
  if (from_len <= avg_refs) {
    remaining_to_move = 0;
  } else {
    remaining_to_move = from_len - avg_refs;
  }
}
```
or more likely
```
size_t remaining_to_move = from_len;
if (from_idx < _num_queues) {
  remaining_to_move -= MIN2(from_len, avg_refs);
}
```
The calculation in line 3 is an ancient and venerable idiom for "subtract,
clipping to non-negative".